File: Classe.java
import java.util.List;
import java.util.ArrayList;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.io.FileReader;
import java.io.IOException;
import java.io.BufferedReader;

class Personagem{
    private String id;
    private String nome;
    private List<String> nome_alternativo;
    private String casa;
    private String ancestral;
    private String especie;
    private String patronus;
    private Boolean staff;
    private String estudante;
    private String nomeator;
    private Boolean vivo;
    private String atoresalternativos;
    private Date dianascimento;
    private int anonascimento;
    private String corolho;
    private String sexo;
    private String corcabelo;
    private Boolean bruxo;

    public Personagem(){
        nome_alternativo = new ArrayList<>();
        this.staff = false;
        this.vivo=false;
        this.bruxo=false;
    }
    // Setters
    public void setId(String id) {
        this.id = id;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public void setCasa(String casa) {
        this.casa = casa;
    }

    public void setNomeAlternativo(List<String> nomeAlternativo) {
        this.nome_alternativo = new ArrayList<>(nomeAlternativo);
    }

    public void addNomeAlternativo(String nomeAlternativo) {
        this.nome_alternativo.add(nomeAlternativo);
    }

    public void setAncestral(String ancestral) {
        this.ancestral = ancestral;
    }

    public void setEspecie(String especie) {
        this.especie = especie;
    }

    public void setPatronus(String patronus) {
        this.patronus = patronus;
    }

    public void setStaff(boolean staff) {
        this.staff = staff;
    }

    public void setEstudante(String estudante) {
        this.estudante = estudante;
    }

    public void setNomeator(String nomeator) {
        this.nomeator = nomeator;
    }

    public void setVivo(boolean vivo) {
        this.vivo = vivo;
    }
    public void setAtorAlternativo(String dados){
            dados = dados.replace("[", "").replace("]", "").replace("'", "").trim();
            if (!dados.isEmpty()) {
                String[] nomes = dados.split(", ");
                for (String nome : nomes) {
                    this.nome_alternativo.add(nome);
                }
            }
    }

    public void setDianascimento(String dianascimentoStr) {
        SimpleDateFormat sdf = new SimpleDateFormat("dd-MM-yyyy");
        try {
            this.dianascimento = sdf.parse(dianascimentoStr);
        } catch (Exception e) {
            System.out.println("Formato de data inválido: " + e.getMessage());
            this.dianascimento = null; 
        }
    }

    public void setAnonascimento(int anonascimento) {
        this.anonascimento = anonascimento;
    }

    public void setCorolho(String corolho) {
        this.corolho = corolho;
    }

    public void setSexo(String sexo) {
        this.sexo = sexo;
    }

    public void setCorcabelo(String corcabelo) {
        this.corcabelo = corcabelo;
    }

    public void setBruxo(boolean bruxo) {
        this.bruxo = bruxo;
    }
// Getters
        public String getId() {
            return id;
        }
    
        public String getNome() {
            return nome;
        }
    
        public String getCasa() {
            return casa;
        }
    
        public List<String> getNomeAlternativo() {
            return new ArrayList<>(nome_alternativo);
        }
    
        public String getAncestral() {
            return ancestral;
        }
    
        public String getEspecie() {
            return especie;
        }
    
        public String getPatronus() {
            return patronus;
        }
    
        public boolean isStaff() {
            return staff;
        }
    
        public String isEstudante() {
            return estudante;
        }
    
        public String getNomeator() {
            return nomeator;
        }
    
        public boolean isVivo() {
            return vivo;
        }
        public String getatoralternativo(){
            return atoresalternativos;
        }
    
        public String getDianascimento() {
            SimpleDateFormat sdf = new SimpleDateFormat("dd-MM-yyyy");
            if (this.dianascimento != null) {
                return sdf.format(this.dianascimento);
            } else {
                return "Data de nascimento não disponível";
            }
            }
    
        public int getAnonascimento() {
            return anonascimento;
        }
    
        public String getCorolho() {
            return corolho;
        }
    
        public String getSexo() {
            return sexo;
        }
    
        public String getCorcabelo() {
            return corcabelo;
        }
    
        public boolean isBruxo() {
            return bruxo;
        }
    

        public void ler(String codigo){
            try (BufferedReader br = new BufferedReader(new FileReader("/tmp/characters.csv"))) {
                String linha;
                String[] campos;
                boolean personagemEncontrado = false;
        
                while ((linha = br.readLine()) != null) {
                    campos = linha.split(";");
                    if (campos[0].equals(codigo)) {
                        setId(campos[0]);
                        setNome(campos[1]);
                        addNomeAlternativo(campos[2]);
                        setCasa(campos[3]);
                        setAncestral(campos[4]);
                        setEspecie(campos[5]);
                        setPatronus(campos[6]);
                        setStaff(Boolean.parseBoolean(campos[7]));
                        setEstudante(campos[8]);
                        setNomeator(campos[9]);
                        setVivo(Boolean.parseBoolean(campos[10]));
                        setAtorAlternativo(campos[11]);
                        setDianascimento(campos[12]);
                        setAnonascimento(Integer.parseInt(campos[13]));
                        setCorolho(campos[14]);
                        setSexo(campos[15]);
                        setCorcabelo(campos[16]);
                        setBruxo(Boolean.parseBoolean(campos[17]));
                        
                        personagemEncontrado = true;
                        break; 
                    }
                }
        
                if (!personagemEncontrado) {
                    System.out.println("Personagem não encontrado.");
                }
            } catch (IOException | NumberFormatException e) {
                System.out.println("Erro ao ler arquivo ou formatar: " + e.getMessage());
            }
        }
public void imprimir(){
    MyIO.println("["+getId()+" ## "+getNome()+" ## "+getNomeAlternativo()+" ## "+getCasa()+" ## "+getAncestral()+" ## "+getEspecie()+" ## "+getPatronus()+" ## "+isStaff()+" ## "+isEstudante()+" ## "+getNomeator()+" ## "+isVivo()+" ## "+getDianascimento()+" ## "+getAnonascimento()+" ## "+getCorolho()+" ## "+getSexo()+" ## "+getCorcabelo()+" ## "+isBruxo()+"]");
}
}


class Classe{
    public static void main(String args []){
        String linha=MyIO.readLine();
        while(!linha.equals("FIM")){
            Personagem personagem = new Personagem();
            personagem.ler(linha);
            personagem.imprimir();
            linha=MyIO.readLine();
            }
    }
}

File: Classe.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>


#define FILE_PATH "/tmp/characters.csv"

typedef struct {
    char *id;
    char *nome;
    char *nome_alternativo;
    char *casa;
    char *ancestral;
    char *especie;
    char *patronus;
    bool staff;
    char *estudante;
    char *nomeator;
    bool vivo;
    char *atoresalternativos;
    char *dianascimento;
    int anonascimento;
    char *corolho;
    char *sexo;
    char *corcabelo;
    bool bruxo;
} Personagem;

void free_string_array(char **array, int length) {
    if (array == NULL)
        return;

    for (int i = 0; i < length; i++)
        free(array[i]);
    free(array);
}

char *read_string(FILE *file) {
    char *str = NULL;
    int capacity = 16; // Initial capacity
    int length = 0;
    char c;

    while ((c = fgetc(file)) != EOF && c != ';') {
        if (length >= capacity - 1) {
            capacity *= 2;
            str = realloc(str, capacity);
            if (str == NULL) {
                perror("Memory allocation error");
                return NULL;
            }
        }
        str[length++] = c;
    }
    str[length] = '\0';

    return str;
}

bool read_boolean(FILE *file) {
    char *str = read_string(file);
    if (str == NULL) {
        perror("Error reading boolean");
        return false;
    }
    bool value = (strcmp(str, "true") == 0);
    free(str);
    return value;
}

char **read_multivalued(FILE *file, int *count) {
    char **array = NULL;
    int capacity = 5; // Initial capacity
    int length = 0;
    char c;

    while ((c = fgetc(file)) != EOF && c != ']') {
        if (c == '[' || c == ' ')
            continue;
        if (length >= capacity) {
            capacity *= 2;
            array = realloc(array, capacity * sizeof(char *));
            if (array == NULL) {
                perror("Memory allocation error");
                return NULL;
            }
        }
        array[length++] = read_string(file);
    }
    *count = length;
    return array;
}

Personagem ler(FILE *file, char *codigo) {
    Personagem p;
    char linha[1024];

    while (fgets(linha, sizeof(linha), file) != NULL) {
        char *separador = strtok(linha, ";");
        p.id = strdup(separador);
        if (strcmp(p.id, codigo) == 0) {
            separador = strtok(NULL, ";");
            p.nome = strdup(separador);
            separador = strtok(NULL, ";");
            p.nome_alternativo = strdup(separador);
            separador = strtok(NULL, ";");
            p.casa = strdup(separador);
            separador = strtok(NULL, ";");
            p.ancestral = strdup(separador);
            separador = strtok(NULL, ";");
            p.especie = strdup(separador);
            separador = strtok(NULL, ";");
            p.patronus = strdup(separador);
            separador = strtok(NULL, ";");
            p.staff = (strcmp(separador, "true") == 0);
            separador = strtok(NULL, ";");
            p.estudante = strdup(separador);
            separador = strtok(NULL, ";");
            p.nomeator = strdup(separador);
            separador = strtok(NULL, ";");
            p.vivo = (strcmp(separador, "true") == 0);
            separador = strtok(NULL, ";");
            p.atoresalternativos = strdup(separador);
            separador = strtok(NULL, ";");
            p.dianascimento = strdup(separador);
            separador = strtok(NULL, ";");
            p.anonascimento = atoi(separador);
            separador = strtok(NULL, ";");
            p.corolho = strdup(separador);
            separador = strtok(NULL, ";");
            p.sexo = strdup(separador);
            separador = strtok(NULL, ";");
            p.corcabelo = strdup(separador);
            separador = strtok(NULL, ";");
            p.bruxo = (strcmp(separador, "true") == 0);
            return p;
        }
    }
    perror("Personagem não encontrado");
    exit(1);
}

void imprimir(Personagem p) {
    printf("[%s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %d ## %s ## %s ## %s ## %s]\n",
           p.id, p.nome, p.nome_alternativo, p.casa, p.ancestral, p.especie, p.patronus,
           p.staff ? "true" : "false", p.estudante, p.nomeator, p.vivo ? "true" : "false",
           p.dianascimento, p.anonascimento, p.corolho, p.sexo, p.corcabelo, p.bruxo ? "true" : "false");
}

int main() {
    FILE *file = fopen(FILE_PATH, "r");
    if (file == NULL) {
        perror("Erro ao abrir o arquivo");
        return 1;
    }

    char codigo[60];
    scanf("%s", codigo);

    Personagem p = ler(file, codigo);
    
    imprimir(p);

    fclose(file);
    return 0;
}
